volatility3/framework/plugins/linux/capabilities.py

```diff
@@ -0,0 +1,207 @@
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import logging
+from dataclasses import dataclass, astuple, fields
+from typing import Iterable, List, Tuple, Dict
+
+from volatility3.framework import interfaces, renderers, exceptions
+from volatility3.framework.constants.linux import CAP_FULL
+from volatility3.framework.configuration import requirements
+from volatility3.framework.interfaces import plugins
+from volatility3.framework.objects import utility
+from volatility3.framework.symbols.linux import extensions
+from volatility3.plugins.linux import pslist
+
+vollog = logging.getLogger(__name__)
+
+
+@dataclass
+class TaskData:
+    """Stores basic information about a task"""
+
+    comm: str
+    pid: int
+    tgid: int
+    ppid: int
+    euid: int
+
+
+@dataclass
+class CapabilitiesData:
+    """Stores each set of capabilties for a task"""
+
+    cap_inheritable: interfaces.objects.ObjectInterface
+    cap_permitted: interfaces.objects.ObjectInterface
+    cap_effective: interfaces.objects.ObjectInterface
+    cap_bset: interfaces.objects.ObjectInterface
+    cap_ambient: interfaces.objects.ObjectInterface
+
+    def astuple(self) -> Tuple:
+        """Returns a shallow copy of the capability sets in a tuple.
+
+        Otherwise, when dataclasses.astuple() performs a deep-copy recursion on
+        ObjectInterface will take a substantial amount of time.
+        """
+        return tuple(getattr(self, field.name) for field in fields(self))
+
+
+class Capabilities(plugins.PluginInterface):
+    """Lists process capabilities"""
+
+    _required_framework_version = (2, 0, 0)
+
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Linux kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.PluginRequirement(
+                name="pslist", plugin=pslist.PsList, version=(2, 0, 0)
+            ),
+            requirements.ListRequirement(
+                name="pids",
+                description="Filter on specific process IDs.",
+                element_type=int,
+                optional=True,
+            ),
+        ]
+
+    def _check_capabilities_support(
+        self,
+        context: interfaces.context.ContextInterface,
+        vmlinux_module_name: str,
+    ):
+        """Checks that the framework supports at least as much capabilities as
+        the kernel being analysed. Otherwise, it shows a warning for the
+        developers.
+        """
+
+        vmlinux = context.modules[vmlinux_module_name]
+
+        try:
+            kernel_cap_last_cap = vmlinux.object_from_symbol(symbol_name="cap_last_cap")
+        except exceptions.SymbolError:
+            # It should be a kernel < 3.2
```

Comment:
Just wanted to check this doesn't need a log message (even just a debug one)?

Reply:
It doesn't need a log there. Because, old kernels didn't have this symbol, so in those cases we are unable to check that if we support all the capabilities. Anyway, it doesn't matter, we need this for the latest kernels. The older kernels we already know the capabilities list, it's something I had to check when I wrote it code.